### DISCUSSION TITLE
Custom directive on object

### DISCUSSION THREAD

**Discussion by dan-f** [graphql-java/graphql-java#3015] [Q&A]
Hi,

I'm having trouble implementing a directive via `SchemaDirectiveWiring`  for object types. E.g.:

```graphql
directive @foo on OBJECT

type Query {
  bar: Bar!
}

type Bar @foo {
  id: ID!
}
```

```kt
class FooDirective : SchemaDirectiveWiring {
  override fun onObject(
    environment: SchemaDirectiveWiringEnvironment<GraphQLObjectType>
  ): GraphQLObjectType {
    // what actually goes in here?
  }
```

I've tried a number of implementations, including what's indicated over here: https://github.com/graphql-java/graphql-java/blob/master/src/test/groovy/graphql/schema/idl/SchemaGeneratorDirectiveHelperTest.groovy#L471

For some reason, my wrapping (outer) data fetcher is just not getting invoked. I don't know if it's significant, but I've noticed that when calling `GraphQLCodeRegistry.dataFetcher`, it's unclear what to actually pass as the parent type vs. field. I.e. I'd have expected some internal representation of `Query` to be the parent type with `bar` as the field, but I can see in the debugger that `environment.element` and `environment.fieldsContainer` both point to the same exact object -- the `GraphQLObjectType` instance representing the `Bar`, with no obvious reference to the parent on any of the `environment` APIs.

The other thing that seems odd to me is the fact that the code sample from the test suite is wrapping the child fields of the object type, rather than that object type itself as a child of it's own parent. I'm hoping there's a straightforward answer to this!

**Comment by bbakerman** [comment]
```
type Bar @foo {
```

The thing is - there is NEVER  a data fetcher on a type - data fetchers ONLY go on fields.

So the question is - what do you want to happen when a "field" contained in `type Bar` is fetched?

if you want to change ALL the data fetchers of fields in `type Bar` then you need to intercept them?

You can do this in `GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment)` and `environment.getFieldsContainer()` will equal `Bar` when you go past the `id` field say

```
           GraphQLFieldDefinition onField(SchemaDirectiveWiringEnvironment<GraphQLFieldDefinition> environment) {
                String name = environment.getElement().getName()
                GraphQLFieldDefinition fieldDefinition = environment.getFieldDefinition()

                GraphQLFieldsContainer fieldsContainer = environment.getFieldsContainer()
                GraphQLNamedType namedType = (GraphQLNamedType) fieldsContainer
                GraphQLDirectiveContainer directiveContainer = (GraphQLDirectiveContainer) fieldsContainer
                // both object types and interfaces can be treated as such
                def fooAppliedDirective = directiveContainer.getAppliedDirective("foo")
                FieldCoordinates coordinates = FieldCoordinates.coordinates(fieldsContainer, fieldDefinition)

                if (fooAppliedDirective != null) {
                    DataFetcher original = environment.getFieldDataFetcher()
                    DataFetcher newDF = new DataFetcher() {
                        @Override
                        Object get(DataFetchingEnvironment dfe) throws Exception {
                            return original.get(dfe)
                        }
                    }

                    environment.getCodeRegistry().dataFetcher(coordinates, newDF)
                }

                return fieldDefinition
            }
```

The above is pure delegation - it grabs the fields OLD data fetcher and creates a new one and delegates back to it.  You can of course introduce conditional code

It looks at the container ( the type which will be an object or interface) and treats it as a directive container which we know it is by java type design.

**Reply by bbakerman** [reply]
sorry its in half java half groovy

**Comment by bbakerman** [comment]
ps.  `graphql.schema.idl.SchemaDirectiveWiring#onObject` is called when all the fields have been iterated - so in theory can can do it per object by putting in new data fetchers for a objects fields
